core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/decorator/FilteringMoveSelector.java

```diff
@@ -68,23 +74,53 @@ public long getSize() {
 
     @Override
     public Iterator<Move<Solution_>> iterator() {
-        return new JustInTimeFilteringMoveIterator(childMoveSelector.iterator(), determineBailOutSize());
+        return new JustInTimeFilteringMoveIterator(childMoveSelector.iterator(), determineBailOutSize(), phaseScope);
+    }
+
+    private long determineBailOutSize() {
+        if (!bailOutEnabled) {
+            return -1L;
+        }
+        try {
+            return childMoveSelector.getSize() * BAIL_OUT_MULTIPLIER;
+        } catch (Exception ex) {
+            // Some move selectors throw an exception when getSize() is called.
+            // In this case, we choose to disregard it and pick a large-enough bail-out size anyway.
+            // The ${bailOutSize+1}th move could in theory show up where previous ${bailOutSize} moves did not,
+            // but we consider this to be an acceptable risk,
+            // outweighed by the benefit of the solver never running into an endless loop.
+            // The exception itself is swallowed, as it doesn't bring any useful information.
+            long bailOutSize = Short.MAX_VALUE * BAIL_OUT_MULTIPLIER;
+            logger.trace(
+                    "        Never-ending move selector ({}) failed to provide size, choosing a bail-out size of ({}) attempts.",
+                    childMoveSelector, bailOutSize);
+            return bailOutSize;
+        }
     }
 
     private class JustInTimeFilteringMoveIterator extends UpcomingSelectionIterator<Move<Solution_>> {
 
         private final Iterator<Move<Solution_>> childMoveIterator;
         private final long bailOutSize;
+        private final AbstractPhaseScope<Solution_> phaseScope;
+        private final PhaseTermination<Solution_> termination;
 
-        public JustInTimeFilteringMoveIterator(Iterator<Move<Solution_>> childMoveIterator, long bailOutSize) {
+        public JustInTimeFilteringMoveIterator(Iterator<Move<Solution_>> childMoveIterator, long bailOutSize,
+                AbstractPhaseScope<Solution_> phaseScope) {
             this.childMoveIterator = childMoveIterator;
             this.bailOutSize = bailOutSize;
+            this.phaseScope = phaseScope;
+            this.termination = phaseScope != null ? phaseScope.getTermination() : null;
         }
 
         @Override
         protected Move<Solution_> createUpcomingSelection() {
             Move<Solution_> next;
             long attemptsBeforeBailOut = bailOutSize;
+            // To reduce the impact of checking for termination on each move,
+            // we only check for termination
+            // after filtering out a total number of moves equal to size(childMoveSelector).
+            long attemptsBeforeCheckTermination = bailOutSize / BAIL_OUT_MULTIPLIER;
```

Comment:
I'm wondering if this bail-out size-based approach isn't too complicated. Why not check for termination every Nth invocation? Say 1000th? 

Otherwise if you have to do millions of iterations before checking the iteration, it will look as if the termination is ignored anyway.